messaging/messaging.go

```diff
@@ -0,0 +1,227 @@
+type WebpushNotification struct {
+	Title string `json:"title,omitempty"`
+	Body  string `json:"body,omitempty"`
+	Icon  string `json:"icon,omitempty"`
+}
+
+// ApnsConfig is Apple Push Notification Service specific options.
+// See https://firebase.google.com/docs/reference/fcm/rest/v1/projects.messages#ApnsConfig
+type ApnsConfig struct {
+	Headers map[string]interface{} `json:"headers,omitempty"`
+	Payload map[string]interface{} `json:"payload,omitempty"`
+}
+
+// NewClient creates a new instance of the Firebase Cloud Messaging Client.
+//
+// This function can only be invoked from within the SDK. Client applications should access the
+// the Messaging service through firebase.App.
+func NewClient(ctx context.Context, c *internal.MessagingConfig) (*Client, error) {
+	if c.ProjectID == "" {
+		return nil, errors.New("project id is required to access firebase cloud messaging client")
+	}
+
+	hc, _, err := transport.NewHTTPClient(ctx, c.Opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		endpoint: messagingEndpoint,
+		client:   &internal.HTTPClient{Client: hc},
+		project:  c.ProjectID,
+		version:  "Go/Admin/" + c.Version,
+	}, nil
+}
+
+// Send sends a Message to Firebase Cloud Messaging.
+//
+// Send a message to specified target (a registration token, topic or condition).
+// https://firebase.google.com/docs/cloud-messaging/send-message
+func (c *Client) Send(ctx context.Context, message *Message) (string, error) {
+	if err := validateMessage(message); err != nil {
+		return "", err
+	}
+	payload := &requestMessage{
+		Message: *message,
+	}
+	return c.sendRequestMessage(ctx, payload)
+}
+
+// SendDryRun sends a dryRun Message to Firebase Cloud Messaging.
+//
+// Send a message to specified target (a registration token, topic or condition).
+// https://firebase.google.com/docs/cloud-messaging/send-message
+func (c *Client) SendDryRun(ctx context.Context, message *Message) (string, error) {
+	if err := validateMessage(message); err != nil {
+		return "", err
+	}
+	payload := &requestMessage{
+		ValidateOnly: true,
+		Message:      *message,
+	}
+	return c.sendRequestMessage(ctx, payload)
+}
+
+func (c *Client) sendRequestMessage(ctx context.Context, payload *requestMessage) (string, error) {
+	versionHeader := internal.WithHeader("X-Client-Version", c.version)
+
+	request := &internal.Request{
+		Method: http.MethodPost,
+		URL:    fmt.Sprintf("%s/projects/%s/messages:send", c.endpoint, c.project),
+		Body:   internal.NewJSONEntity(payload),
+		Opts:   []internal.HTTPOption{versionHeader},
+	}
+	resp, err := c.client.Do(ctx, request)
+	if err != nil {
+		return "", err
+	}
+
+	if _, ok := errorCodes[resp.Status]; ok {
+		return "", fmt.Errorf("unexpected http status code : %d, reason: %v", resp.Status, string(resp.Body))
+	}
+
+	result := &ResponseMessage{}
+	err = resp.Unmarshal(http.StatusOK, result)
+
+	return result.Name, err
+}
+
+// validators
+
+// TODO add validator : Data messages can have a 4KB maximum payload.
```

Comment:
In other SDKs I've implemented following validations:

* Topic name doesn't have the `/topics/` prefix.
* Topic name contains the valid
* Exactly one of topic, token or condition is specified
* AndroidConfig.Ttl has the correct format
* AndroidConfig.Priority value is valid
* AndroidNotification.Color has the correct format
* TitleLocKey is specified when TitleLocArgs are specified
* BodyLocKey is specified when BodyLocArgs are specified

See https://github.com/firebase/firebase-admin-java/pull/127 for reference.